lib/providers/calendar_data.dart

```diff
@@ -0,0 +1,105 @@
+import 'dart:convert';
+
+import 'package:concordia_navigation/models/schedule.dart';
+import 'package:concordia_navigation/services/network.dart';
+import 'package:concordia_navigation/storage/app_constants.dart' as constants;
+import 'package:device_calendar/device_calendar.dart';
+import 'package:flutter/widgets.dart';
+
+class CalendarData extends ChangeNotifier {
+  final DeviceCalendarPlugin _deviceCalendarPlugin = DeviceCalendarPlugin();
+  List<Calendar> _calendars = List();
+  List<Event> _classes = List();
+  Schedule _schedule;
+
+  Schedule get schedule => _schedule;
+
+  CalendarData() {
+    retrieveFromDevice();
+  }
+
+  /// Checks for (and requests) permissions required by the [device calendar plugins](https://pub.dev/packages/device_calendar)
+  Future<bool> _checkPermission() async {
+    var permissionsGranted = await _deviceCalendarPlugin.hasPermissions();
+    if (permissionsGranted.isSuccess && !permissionsGranted.data) {
+      permissionsGranted = await _deviceCalendarPlugin.requestPermissions();
+      if (!permissionsGranted.isSuccess || !permissionsGranted.data) {
+        throw new Exception('Missing Calendar Permissions');
+      }
+    }
+    return true;
+  }
+
+  /// Retrieves calendar objects from the device that match a certain format.
+  Future<void> _retrieveDeviceCalendars() async {
```

Comment:
Whenever a new event is added or modified in the calendar, the change is not reflected in the app, unless the app is re-launched. 
One Solution is to add a refresh button in the Appbar to re-retrieve the data from the calendar.

Reply:
agreed. However, that refresh button should call the calendar provider@`retrieveFromDevice()` instead of the private `_retrieveDeviceCalendars()`. That method will check the already existing calendars for new events, update the schedule object and notify widgets. 

If we need to, we can remove the calendar empty check and just always reload calendars before fetching events.

Reply:
I added a refresh button and tested it on IOS and Android.

Reply:
resolved in eefe3708da0bb67ff9273b8fadb60609466f6caf